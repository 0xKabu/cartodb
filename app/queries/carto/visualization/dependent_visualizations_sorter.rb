# encoding: UTF-8

require 'active_record'
require 'postgres_ext'

class Carto::VisualizationQueryFilterer

  def initialize(query: Carto::Visualization.all)
    @query = query
  end

  def sort
    @query = query.with(dependencies: dependencies_query)
    @query = query.select('coalesce(dependent_visualizations_count, 0)')
    @query = query.joins('LEFT OUTER JOIN dependencies ON dependencies.id = visualizations.id')
    # TODO: add filters
    # TODO: sort
  end

  private

  def dependencies_query
    nested_association = { map: { user_table: { layers: { maps: :visualization } } } }
    select_count = 'count(distinct(visualizations.id, visualizations_maps.id)) as dependent_visualizations_count'
    Carto::Visualization.select(:id, select_count)
                        .joins(nested_association)
                        .where(visualizations_maps: { type: 'derived' })
                        .group(:id)
  end

end
