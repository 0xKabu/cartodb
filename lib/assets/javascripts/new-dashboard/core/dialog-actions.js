// Dialogs
import Dialog from 'new-dashboard/components/Backbone/Dialog.vue';
import ChangePrivacy from 'new-dashboard/components/Backbone/Dialogs/ChangePrivacy';
import DuplicateMap from 'new-dashboard/components/Backbone/Dialogs/DuplicateMap';
import LockMap from 'new-dashboard/components/Backbone/Dialogs/LockMap';
import DeleteDialog from 'new-dashboard/components/Backbone/Dialogs/DeleteDialog';

export function changePrivacy (visualization) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <ChangePrivacy :visualization="visualization" v-on:close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, ChangePrivacy }
    },
    { visualization }
  );
}

export function duplicateMap (visualization) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <DuplicateMap :visualization="visualization" v-on:close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, DuplicateMap }
    },
    { visualization }
  );
}

export function changeLockState (visualization, contentType) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <LockMap :visualization="visualization" :contentType="contentType" v-on:close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, LockMap }
    }, dialogProps
  );
}

export function changeVisualizationsLockState (visualizations, contentType) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <LockMap :visualizations="visualizations" :contentType="contentType" v-on:close="$emit('close')"/>
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, LockMap }
    }, dialogProps
  );
}

export function deleteVisualization (visualization, contentType) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <DeleteDialog :visualization="visualization" :contentType="contentType" v-on:close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, DeleteDialog }
    }, dialogProps
  );
}

export function deleteVisualizations (visualizations, contentType) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <DeleteDialog :visualizations="visualizations" :contentType="contentType" v-on:close="$emit('close')"/>
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, DeleteDialog }
    }, dialogProps
  );
}
