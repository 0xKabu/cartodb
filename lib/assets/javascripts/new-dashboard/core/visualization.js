import PermissionModel from 'dashboard/data/permission-model';
import store from 'new-dashboard/store';

export function getURL (visualizationData) {
  const loggedUserBackboneModel = store.state.user.userModel;
  const permissionModel = new PermissionModel(
    visualizationData.permission, { configModel: store.state.config.configModel }
  );

  let id = visualizationData.id;
  let userUrl = permissionModel.owner.viewUrl();

  if (loggedUserBackboneModel &&
      loggedUserBackboneModel.id !== permissionModel.owner.get('id') &&
      permissionModel.hasAccess(loggedUserBackboneModel)) {
    userUrl = loggedUserBackboneModel.viewUrl();
    id = `${permissionModel.owner.get('username')}.${id}`;
  }

  return `${userUrl.toString()}/viz/${id}`;
};
