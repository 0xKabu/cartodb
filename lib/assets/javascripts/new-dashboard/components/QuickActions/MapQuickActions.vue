<template>
  <QuickActions :actions="actions" v-on="getEventListeners()"></QuickActions>
</template>

<script>
import QuickActions from 'new-dashboard/components/QuickActions/QuickActions';
import * as DialogActions from 'new-dashboard/core/dialog-actions';

export default {
  name: 'MapQuickActions',
  components: {
    QuickActions
  },
  props: {
    map: Object
  },
  data () {
    return {
      actions: [
        { name: this.$t('QuickActions.maps.editInfo'), event: 'editInfo' },
        { name: this.$t('QuickActions.maps.changePrivacy'), event: 'changePrivacy' },
        { name: this.$t('QuickActions.maps.manageTags'), event: 'manageTags' },
        { name: this.$t('QuickActions.maps.duplicate'), event: 'duplicateMap' },
        { name: this.$t('QuickActions.maps.lock'), event: 'lockMap' },
        { name: this.$t('QuickActions.maps.delete'), event: 'deleteMap', isDestructive: true }
      ]
    };
  },
  methods: {
    getEventListeners () {
      const events = this.actions.map(action => action.event);

      console.log(events);
      return events.reduce(
        (eventListeners, action) => {
          eventListeners[action] = this[action].bind(this);
          return eventListeners;
        }, {}
      );
    },
    showModal (componentDefinition, componentPropsData) {
      this.$modal.show(
        componentDefinition,
        componentPropsData,
        { width: '100%', height: '100%' }
      );
    },
    editInfo () {
      DialogActions.editMapMetadata.apply(this, [this.map]);
    },
    changePrivacy () {
      DialogActions.changePrivacy.apply(this, [this.map]);
    },
    manageTags () {
      DialogActions.editMapMetadata.apply(this, [this.map]);
    },
    duplicateMap () {
      DialogActions.duplicateMap.apply(this, [this.map]);
    },
    unlockMap () {
      DialogActions.changeLockState.apply(this, [this.map, 'maps']);
    },
    lockMap () {
      DialogActions.changeLockState.apply(this, [this.map, 'maps']);
    },
    deleteMap () {
      DialogActions.deleteVisualization.apply(this, [this.map, 'maps']);
    }
  }
};
</script>
