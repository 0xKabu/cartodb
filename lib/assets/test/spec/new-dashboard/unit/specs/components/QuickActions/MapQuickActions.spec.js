import { shallowMount } from '@vue/test-utils';
import MapQuickActions from 'new-dashboard/components/QuickActions/MapQuickActions';
import visualization from '../../fixtures/visualizations';
import * as DialogActions from '../__mocks__/dialog-actions';

const $t = key => key;

describe('MapQuickAction.vue', () => {
  it('should render correct contents', () => {
    const map = visualization.visualizations[0];
    const mapQuickAction = shallowMount(MapQuickActions, {
      propsData: { map },
      mocks: {
        $t
      }
    });

    expect(mapQuickAction).toMatchSnapshot();
  });

  describe('methods', () => {
    it('should open edit metadata modal', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $t
        }
      });

      mapQuickAction.vm.editInfo();

      expect(DialogActions.editMapMetadata).toHaveBeenCalled();
    });

    it('should open change privacy modal', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $t
        }
      });

      mapQuickAction.vm.changePrivacy();

      expect(DialogActions.changePrivacy).toHaveBeenCalled();
    });

    it('should open manage tags modal', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $t
        }
      });

      mapQuickAction.vm.manageTags();

      expect(DialogActions.editMapMetadata).toHaveBeenCalled();
    });

    it('should launch duplicate map action', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $t
        }
      });

      mapQuickAction.vm.duplicateMap();

      expect(DialogActions.duplicateMap).toHaveBeenCalled();
    });

    it('should launch unlock map action', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $t
        }
      });

      mapQuickAction.vm.unlockMap();

      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should launch lock map action', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $t
        }
      });

      mapQuickAction.vm.lockMap();

      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should launch delete map action', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $t
        }
      });

      mapQuickAction.vm.deleteMap();

      expect(DialogActions.deleteVisualization).toHaveBeenCalled();
    });
  });
});
