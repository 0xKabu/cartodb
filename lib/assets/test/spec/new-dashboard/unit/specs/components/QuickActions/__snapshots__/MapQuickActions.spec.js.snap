// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapQuickAction.vue should render correct contents 1`] = `<quickactions-stub actions="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"></quickactions-stub>`;
